substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/image/NativeImageDebugInfoProvider.java

```diff
@@ -879,6 +914,22 @@ private ResolvedJavaMethod promoteAnalysisToHosted(ResolvedJavaMethod m) {
             return m;
         }
 
+        private ResolvedJavaMethod originalMethod() {
```

Comment:
No, it's not exactly the same and the difference is very important.

This method is used to compute the /identity/ for a method presented anywhere in the DebugInfoProvider API. It unwraps from the Hosted domain to the JVMCI domain so that we ensure we get the same method whether the NativeImage or CodeCache presents it as a Hosted value or an original JVMCI value.

In the one special case of a `SubstitutionMethod` we traverse to the /original/ method. That ensures we consistently identify a substitution `Bar.bar` <--@Sub--  `Foo.foo`  with method `foo`, not method `bar`. We need to do that so that in the type info model we only have the one method `foo` declared in the one class `Foo`. The debugger has to see a single, consistent view of the class and method. We don't want to see a method called `bar` attached to class `Foo` in one context and a method called `foo`  attached to `Foo` in another context (and we certainly don't want to see a method called `bar` attached to class `Bar`). In fact we don't want to see class `Bar` referenced at all in the debug info type model (although we do want to see the file and source code for `Bar.bar` when debugging).

The only other case where we don't unwrap to a JVMCI method is when we hit some other form of substitution. In that case we use the unwrapped method is our id method. We don't ever traverse to the original here even when, in some cases, we can do so. That's for two reasons. Firstly, for custom substitutions the original is not normally of any relevance. The method code is synthesized, which means the original's file and bytecode is irrelevant. Secondly, the original may actually exist as a method in its own right, listed as belonging to the declaring class (this happens with custom substitution cases for foreign stub methods and reflective methods). If we used the original to identify the substitution we would end up conflating two or more independent methods (the debug model type scan detects this, generating an error because the same method is notified twice).

Method `getOriginal(HostedMethod)` (now renamed as `getAnnotatedOrOriginal(HostedMethod)`) is only used when determining a method's file or permissions. It also unwraps from the Hosted domain to the JVMCI domain. The difference is that in the special case of a `SubstitutionMethod` it traverses to the /annotated/ method. That's needed for file lookup because the correct file needs to be derived from the declaring class of the annotated method. It's also necessary as far as permissions are concerned because the compiler compiles the bytecode from the  /annotated/ method so generates code whose permissions are determined by that bytecode.

Reply:
Thanks for the clarification Andrew. I missed the difference in my first pass.